File: api/src/dressme/background_tasks.py
import logging
from uuid import UUID, uuid4

from sqlmodel import Session, select

from . import db
from .avatar_generation import AvatarGenerator
from .blob_storage import BlobStorage
from .settings import get_settings
from .woa_generation import WoaGenerator

settings = get_settings()


# TODO: add a test for this
def generate_avatar_task(
    *, user_id: UUID, avatar_generator: AvatarGenerator, blob_storage: BlobStorage
):
    """Background task: generates a game avatar from the user's selfie."""
    with Session(db.engine) as session:
        user = session.exec(select(db.User).where(db.User.id == user_id)).one()

        if user.selfie_image_key is None:
            raise ValueError("User does not have a selfie image")

        try:
            selfie_data = blob_storage.download(
                settings.SELFIES_BUCKET, user.selfie_image_key
            )
            avatar_data = avatar_generator.generate(selfie_data)

            avatar_key = f"{uuid4()}.jpg"
            blob_storage.upload(
                settings.AVATARS_BUCKET, avatar_key, avatar_data, "image/jpeg"
            )

            user.avatar_image_key = avatar_key
            session.commit()
            logging.info(f"Avatar generation completed for user {user_id}")
        except Exception:
            logging.exception(f"Avatar generation failed for user {user_id}")


# TODO: add a test for this
async def generate_woa_image_task(
    *,
    wearable_id: UUID,
    user_id: UUID,
    woa_generator: WoaGenerator,
    blob_storage: BlobStorage,
):
    """
    Generates an image of the given user's avatar wearing a given wearable.

    This image is called a WearableOnAvatar (WOA) image. It is generated using AI models.
    This method is intended to be used as a FastAPI background task.
    """
    with Session(db.engine) as session:
        logging.info("Starting WOA generation")
        user = session.exec(select(db.User).where(db.User.id == user_id)).one()
        wearable = session.exec(
            select(db.Wearable)
            .where(db.Wearable.id == wearable_id)
            .where(db.Wearable.user_id == user_id)
        ).one()

        # Generate an image of the avatar wearing the wearable
        logging.info("Generating WOA image")
        if user.avatar_image_key is None:
            raise ValueError("User does not have an avatar image")

        wearable_image_data = blob_storage.download(
            settings.WEARABLES_BUCKET, wearable.image_key
        )
        avatar_image_data = blob_storage.download(
            settings.AVATARS_BUCKET, user.avatar_image_key
        )

        woa_image_data = await woa_generator.generate_image(
            avatar_image=avatar_image_data,
            wearable_image=wearable_image_data,
            wearable_description=wearable.description or "",
            category=wearable.category,
        )

        # Get a mask of the wearable on the avatar using an image segmentation model
        logging.info("Generating mask")
        mask_image_data = await woa_generator.generate_mask(
            woa_image=woa_image_data,
            wearable_description=wearable.description or "",
        )

        # Upload results to blob storage
        logging.info("Uploading results to blob storage")
        woa_key = f"{uuid4()}.jpg"
        mask_key = f"{uuid4()}.jpg"

        blob_storage.upload(settings.WOA_BUCKET, woa_key, woa_image_data, "image/jpeg")
        blob_storage.upload(settings.WOA_BUCKET, mask_key, mask_image_data, "image/jpeg")

        logging.info("Saving results to DB")
        woa_image = db.WearableOnAvatarImage(
            user_id=user.id,
            avatar_image_key=user.avatar_image_key,
            wearable_image_key=wearable.image_key,
            image_key=woa_key,
            mask_image_key=mask_key,
        )
        session.add(woa_image)
        session.commit()
        logging.info("Finished generating WOA image")

File: src/server/background-tasks.ts
import { randomUUID } from "node:crypto";
import { eq } from "drizzle-orm";
import { downloadBlob, uploadBlob } from "./blob-storage";
import { getSettings } from "./settings";
import { db, schema } from "./db";

// TODO: add a test for this
export async function generateAvatarTask(userId: string): Promise<void> {
  const settings = getSettings();

  try {
    const user = await db.query.users.findFirst({
      where: eq(schema.users.id, userId),
    });

    if (!user || !user.selfieImageKey) {
      throw new Error("User does not have a selfie image");
    }

    const selfieData = await downloadBlob(settings.SELFIES_BUCKET, user.selfieImageKey);

    const { generateAvatar } = await import("./avatar-generation");
    const avatarData = await generateAvatar(selfieData);

    const avatarKey = `${randomUUID()}.jpg`;
    await uploadBlob(settings.AVATARS_BUCKET, avatarKey, avatarData, "image/jpeg");

    await db
      .update(schema.users)
      .set({ avatarImageKey: avatarKey })
      .where(eq(schema.users.id, userId));

    console.info(`Avatar generation succeeded for user '${userId}'`);
  } catch (error) {
    console.error(`Avatar generation failed for user '${userId}'`, error);
  }
}

// TODO: add a test for this
export async function generateWoaTask(wearableId: string, userId: string): Promise<void> {
  const settings = getSettings();

  try {
    const user = await db.query.users.findFirst({
      where: eq(schema.users.id, userId),
    });
    if (!user) throw new Error(`User '${userId}' not found`);
    if (!user.avatarImageKey) {
      throw new Error(`User '${userId}' does not have an avatar image`);
    }

    const wearable = await db.query.wearables.findFirst({
      where: eq(schema.wearables.id, wearableId),
    });
    if (!wearable) throw new Error(`Wearable '${wearableId}' not found`);

    const wearableImageData = await downloadBlob(settings.WEARABLES_BUCKET, wearable.imageKey);
    const avatarImageData = await downloadBlob(settings.AVATARS_BUCKET, user.avatarImageKey);

    const { generateWoaImage, generateMask } = await import("./woa-generation");

    // Generate an image of the avatar wearing the wearable
    const woaImageData = await generateWoaImage({
      avatarImage: avatarImageData,
      wearableImage: wearableImageData,
      category: wearable.category,
    });

    // Get a mask of the wearable on the avatar using an image segmentation model
    const maskImageData = await generateMask({
      woaImage: woaImageData,
      category: wearable.category,
    });

    const woaKey = `${randomUUID()}.jpg`;
    const maskKey = `${randomUUID()}.jpg`;

    // Upload results to blob storage
    await uploadBlob(settings.WOA_BUCKET, woaKey, woaImageData, "image/jpeg");
    await uploadBlob(settings.WOA_BUCKET, maskKey, maskImageData, "image/jpeg");

    await db.insert(schema.wearableOnAvatarImages).values({
      userId: user.id,
      avatarImageKey: user.avatarImageKey,
      wearableImageKey: wearable.imageKey,
      imageKey: woaKey,
      maskImageKey: maskKey,
    });

    console.info(`WOA generation succeeded for wearable '${wearableId}'`);
  } catch (error) {
    console.error(`WOA generation failed for wearable '${wearableId}'`, error);
  }
}
